File: src/main/java/com/server/ordering/api/CustomerApiController.java
package com.server.ordering.api;

import com.server.ordering.domain.*;
import com.server.ordering.domain.dto.*;
import com.server.ordering.domain.dto.request.*;
import com.server.ordering.domain.dto.response.*;
import com.server.ordering.domain.member.Customer;
import com.server.ordering.service.CouponService;
import com.server.ordering.service.CustomerService;
import com.server.ordering.service.VerificationService;
import com.server.ordering.service.WaitingService;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.stream.Collectors;

@RestController
@RequiredArgsConstructor
public class CustomerApiController {

    private final VerificationService verificationService;
    private final CustomerService customerService;
    private final CouponService couponService;
    private final WaitingService waitingService;

    /**
     * 인증번호 받기
     */
    @PostMapping("/api/customer/verification/get")
    public ResultDto<Boolean> sendCode(@RequestBody PhoneNumberDto dto) {

        boolean bDuplicatedNumber = verificationService.isPhoneNumberDuplicated(MemberType.CUSTOMER, dto.getPhoneNumber());
        if (bDuplicatedNumber) {
            return new ResultDto<>(1, false);
        }

        verificationService.sendCode(dto.getPhoneNumber());
        return new ResultDto<>(1, true);
    }

    /**
     * 인증번호 체크
     */
    @PostMapping("/api/customer/verification/check")
    public ResultDto<Boolean> checkCode(@RequestBody VerificationDto dto) {
        boolean isSuccess = verificationService.checkCode(dto.getTotalNum(), dto.getCode());
        return new ResultDto<>(1, isSuccess);
    }

    /**
     * 고객 회원 가입
     */
    @PostMapping("/api/customer/signup")
    public ResultDto<Long> singUp(@RequestBody CustomerSignUpDto dto) {
        boolean bIdDuplicated = customerService.isIdDuplicated(dto.getSignInId());
        if (bIdDuplicated) {
            return new ResultDto<>(1, null);
        } else {
            Long customerId = customerService.signUp(dto);
            return new ResultDto<>(1, customerId);
        }
    }

    /**
     * 고객 로그인
     */
    @PostMapping("/api/customer/signin")
    public ResultDto<CustomerSignInResultDto> signIn(@RequestBody SignInDto dto) {
        Customer customer = customerService.signIn(dto);

        if (customer == null) {
            return new ResultDto<>(1, null);
        }

        int basketCount = customerService.getBasketCount(customer.getId());
        return new ResultDto<>(1, new CustomerSignInResultDto(customer.getId(), customer.getNickname(), basketCount));
    }

    /**
     * 고객 로그아웃
     */
    @PostMapping("/api/customer/{customerId}/sign_out")
    public ResultDto<Boolean> signOut(@PathVariable Long customerId) {
        customerService.signOut(customerId);
        return new ResultDto<>(1, true);
    }

    /**
     * 고객 휴대폰번호 변경
     */
    @PutMapping("/api/customer/{customerId}/phone_number")
    public ResultDto<Boolean> putPhoneNumber(@PathVariable Long customerId, @RequestBody PhoneNumberDto dto) {
        customerService.putPhoneNumber(customerId, dto.getPhoneNumber());
        return new ResultDto<>(1, true);
    }

    /**
     * 고객 비밀번호 변경
     */
    @PutMapping("/api/customer/{customerId}/password")
    public ResultDto<Boolean> putPassword(@PathVariable Long customerId, @RequestBody PasswordChangeDto dto) {
        Boolean isChanged = customerService.putPassword(customerId, dto);
        return new ResultDto<>(1, isChanged);
    }

    /**
     * 고객 계정 삭제
     */
    @DeleteMapping("/api/customer/{customerId}")
    public ResultDto<Boolean> delete(@PathVariable Long customerId) {
        customerService.deleteAccount(customerId);
        return new ResultDto<>(1, true);
    }

    /**
     * 고객 리뷰 등록
     */
    @PostMapping("/api/customer/review")
    public ResultDto<Boolean> registerReview(@RequestParam(name = "restaurant_id") Long restaurantId,
                                             @RequestParam(name = "order_id") Long orderId,
                                             @RequestBody ReviewDto dto) {
        Boolean isRegistered = customerService.registerReview(restaurantId, orderId, dto);
        return new ResultDto<>(1, isRegistered);
    }

    /**
     * 고객 리뷰 수정
     */
    @PutMapping("/api/customer/review/{reviewId}")
    public ResultDto<Boolean> putReview(@PathVariable Long reviewId, @RequestBody ReviewDto dto) {
        customerService.putReview(reviewId, dto);
        return new ResultDto<>(1, true);
    }

    /**
     * 고객 리뷰 삭제
     */
    @DeleteMapping("/api/customer/review/{reviewId}")
    public ResultDto<Boolean> putReview(@PathVariable Long reviewId) {
        customerService.removeReview(reviewId);
        return new ResultDto<>(1, true);
    }

    /**
     * 고객 쿠폰 등록
     */
    @PostMapping("/api/customer/{customerId}/coupon")
    public ResultDto<Boolean> getCoupon(@PathVariable Long customerId, @RequestBody CouponSerialNumberDto dto) {
        Coupon coupon = couponService.getCoupon(dto.getSerialNumber());
        Boolean haveCoupon = couponService.haveThisCoupon(dto.getSerialNumber(), customerId);
        if (haveCoupon) {
            return new ResultDto<>(1, false);
        }
        couponService.saveMyCoupon(coupon, customerId);
        return new ResultDto<>(1, true);
    }

    /**
     * 고객 쿠폰 사용
     */
    @DeleteMapping("/api/customer/coupon/{couponId}")
    public ResultDto<Boolean> useCoupon(@PathVariable Long couponId) {
        couponService.useMyCoupon(couponId);
        return new ResultDto<>(1, true);
    }

    /**
     * 내 웨이팅 조회
     */
    @PostMapping("/api/customer/{customerId}/waiting")
    public ResultDto<MyWaitingInfoDto> getMyWaitingInfo(@PathVariable Long customerId) {

        Waiting waiting = waitingService.findOneWithRestaurantByCustomer(customerId);

        // 접수한 웨이팅이 없다면 Null 반환
        if (waiting == null) {
            return new ResultDto<>(1, null);
        }

        // 앞에 남은 사람 수
        long numberInFrontOfMe = waitingService.
                getNumberInFrontOfMe(waiting.getMyWaitingNumber(), waiting.getRestaurant().getId());

        MyWaitingInfoDto myWaitingDto = new MyWaitingInfoDto(waiting, numberInFrontOfMe);
        return new ResultDto<>(1, myWaitingDto);
    }

    /**
     * 장바구니 리스트 조회
     */
    @PostMapping("/api/customer/{customerId}/baskets")
    public ResultDto<List<BasketFood>> getBasketList(@PathVariable Long customerId) {
        List<Basket> baskets = customerService.findBasketWithFood(customerId);
        List<BasketFood> basketResponseDtos = baskets.stream().map(BasketFood::new).collect(Collectors.toList());
        return new ResultDto<>(basketResponseDtos.size(), basketResponseDtos);
    }

    /**
     * 내 쿠폰 리스트 조회
     */
    @PostMapping("/api/customer/{customerId}/my_coupons")
    public ResultDto<List<CouponDto>> getMyCouponList(@PathVariable Long customerId) {
        List<MyCoupon> myCoupons = customerService.findMyCoupon(customerId);
        List<CouponDto> couponDtos = myCoupons.stream().map(CouponDto::new).collect(Collectors.toList());
        return new ResultDto<>(couponDtos.size(), couponDtos);
    }

    /**
     * 매장 북마크(찜)
     */
    @PostMapping("/api/customer/{customerId}/bookmark")
    public ResultDto<Long> restaurantBookmark(@PathVariable Long customerId,
                                              @RequestParam(name = "restaurant_id") Long restaurantId) {
        // 중복 체크
        Boolean bExist = customerService.isExistedBookmark(customerId, restaurantId);
        // 이미 북마크 되어 있음
        if (bExist) {
            return new ResultDto<>(1, null);
        }

        Long bookmarkId = customerService.addBookmark(customerId, restaurantId);
        return new ResultDto<>(1, bookmarkId);
    }

    /**
     * 매장 북마크(찜) 해제
     */
    @DeleteMapping("/api/customer/bookmark/{bookmarkId}")
    public ResultDto<Boolean> deleteBookmark(@PathVariable Long bookmarkId) {
        customerService.removeBookmark(bookmarkId);
        return new ResultDto<>(1, true);
    }

    /**
     * 내 북마크 불러오기
     */
    @GetMapping("/api/customer/{customerId}/bookmarks")
    public ResultDto<List<BookmarkPreviewDto>> getBookmarkList(@PathVariable Long customerId) {
        List<Bookmark> bookmarks = customerService.findAllBookmarkWithRestWithRepresentative(customerId);
        List<BookmarkPreviewDto> previewDtos = bookmarks.stream().map(BookmarkPreviewDto::new).collect(Collectors.toList());
        return new ResultDto<>(previewDtos.size(), previewDtos);
    }

    /**
     * 내 주문 내역(진행 중) 리스트 불러오기
     */
    @GetMapping("/api/customer/{customerId}/orders/ongoing")
    public ResultDto<List<PreviousHistoryDto>> getAllMyOngoingOrders(@PathVariable Long customerId) {
        List<Order> ongoingOrders = customerService.findAllMyOngoingOrders(customerId);
        List<PreviousHistoryDto> historyDtos = ongoingOrders.stream().map(PreviousHistoryDto::new).collect(Collectors.toList());
        return new ResultDto<>(historyDtos.size(), historyDtos);
    }

    /**
     * 내 주문 내역(완료) 리스트 불러오기
     */
    @GetMapping("/api/customer/{customerId}/orders/finished")
    public ResultDto<List<PreviousHistoryDto>> getAllMyFinishedOrders(@PathVariable Long customerId) {
        List<Order> finishedOrders = customerService.findAllMyFinishedOrders(customerId);
        List<PreviousHistoryDto> historyDtos = finishedOrders.stream().map(PreviousHistoryDto::new).collect(Collectors.toList());
        return new ResultDto<>(historyDtos.size(), historyDtos);
    }
}

File: src/main/kotlin/com/byeoru/ordering_server/api/CustomerApiController.kt
package com.byeoru.ordering_server.api

import com.byeoru.ordering_server.domain.*
import com.byeoru.ordering_server.domain.dto.ResultDto
import com.byeoru.ordering_server.domain.dto.request.*
import com.byeoru.ordering_server.domain.dto.response.*
import com.byeoru.ordering_server.service.*
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile


@RestController
class CustomerApiController(val verificationService: VerificationService,
                            val customerService: CustomerService,
                            val couponService: CouponService,
                            val waitingService: WaitingService,
                            val orderService: OrderService,
                            val restaurantService: RestaurantService) {

    /**
     * 인증번호 받기
     */
    @PostMapping("/api/customer/verification/get")
    fun sendCode(@RequestBody dto: PhoneNumberDto): ResultDto<Boolean> {
        val bDuplicatedNumber: Boolean = verificationService.isPhoneNumberDuplicated(MemberType.CUSTOMER, dto.phoneNumber)
        return if (bDuplicatedNumber) {
            ResultDto(1, false)
        } else {
            verificationService.sendCode(dto.phoneNumber)
            ResultDto(1, true)
        }
    }

    /**
     * 인증번호 체크
     */
    @PostMapping("/api/customer/verification/check")
    fun checkCode(@RequestBody dto: VerificationDto): ResultDto<Boolean> {
        val isSuccess: Boolean = verificationService.checkCode(dto.totalNum, dto.code)
        return ResultDto(1, isSuccess)
    }

    /**
     * 고객 회원 가입
     */
    @PostMapping("/api/customer/signup")
    fun singUp(@RequestBody dto: CustomerSignUpDto): ResultDto<Long> {
        val bIdDuplicated: Boolean = customerService.isIdDuplicated(dto.signInId)
        return if (bIdDuplicated) {
            ResultDto(1, null)
        } else {
            val customerId: Long? = customerService.signUp(dto)
            ResultDto(1, customerId)
        }
    }

    /**
     * 고객 로그인
     */
    @PostMapping("/api/customer/signin")
    fun signIn(@RequestBody dto: SignInDto): ResultDto<CustomerSignInResultDto> {
        val customer = customerService.signIn(dto) ?: return ResultDto(1, null)
        val basketCount = customerService.getBasketCount(customer.id!!)
        return ResultDto(1, CustomerSignInResultDto(customer.id!!, customer.nickname, basketCount))
    }

    /**
     * 고객 로그아웃
     */
    @PostMapping("/api/customer/{customerId}/sign_out")
    fun signOut(@PathVariable customerId: Long): ResultDto<Boolean> {
        customerService.signOut(customerId)
        return ResultDto(1, true)
    }

    /**
     * 고객 휴대폰번호 변경
     */
    @PutMapping("/api/customer/{customerId}/phone_number")
    fun putPhoneNumber(@PathVariable customerId: Long,
                       @RequestBody dto: PhoneNumberDto): ResultDto<Boolean> {
        customerService.putPhoneNumber(customerId, dto.phoneNumber)
        return ResultDto(1, true)
    }

    /**
     * 고객 비밀번호 변경
     */
    @PutMapping("/api/customer/{customerId}/password")
    fun putPassword(@PathVariable customerId: Long,
                    @RequestBody dto: PasswordChangeDto): ResultDto<Boolean> {
        val isChanged = customerService.putPassword(customerId, dto)
        return ResultDto(1, isChanged)
    }

    /**
     * 고객 계정 삭제
     */
    @DeleteMapping("/api/customer/{customerId}")
    fun delete(@PathVariable customerId: Long): ResultDto<Boolean> {
        customerService.deleteAccount(customerId)
        return ResultDto(1, true)
    }

    /**
     * 고객 리뷰 등록
     */
    @PostMapping("/api/customer/review")
    fun registerReview(@RequestParam(name = "restaurant_id") restaurantId: Long,
                       @RequestParam(name = "order_id") orderId: Long,
                       @RequestPart(required = false) image: MultipartFile?,
                       @RequestPart dto: ReviewDto): ResultDto<Boolean> {
        val isRegistered = customerService.registerReview(restaurantId, orderId, dto, image)
        return ResultDto(1, isRegistered)
    }

    /**
     * 고객 리뷰 삭제
     */
    @DeleteMapping("/api/customer/review/{reviewId}")
    fun putReview(@PathVariable reviewId: Long): ResultDto<Boolean> {
        customerService.removeReview(reviewId)
        return ResultDto(1, true)
    }

    /**
     * 고객 쿠폰 등록
     */
    @PostMapping("/api/customer/{customerId}/coupon")
    fun getCoupon(@PathVariable customerId: Long,
                  @RequestBody dto: CouponSerialNumberDto): ResultDto<Boolean> {
        val coupon = couponService.getCoupon(dto.serialNumber)
        val haveCoupon = couponService.haveThisCoupon(dto.serialNumber, customerId)
        return if (haveCoupon) {
            ResultDto(1, false)
        } else {
            couponService.saveMyCoupon(coupon, customerId)
            ResultDto(1, true)
        }
    }

    /**
     * 고객 쿠폰 사용
     */
    @DeleteMapping("/api/customer/coupon/{couponId}")
    fun useCoupon(@PathVariable couponId: Long): ResultDto<Boolean> {
        couponService.useMyCoupon(couponId)
        return ResultDto(1, true)
    }

    /**
     * 내 웨이팅 조회
     */
    @PostMapping("/api/customer/{customerId}/waiting")
    fun getMyWaitingInfo(@PathVariable customerId: Long): ResultDto<MyWaitingInfoDto> {
        // 접수한 웨이팅이 없다면 Null 반환
        val waiting = waitingService.findOneWithRestaurantByCustomer(customerId) ?: return ResultDto(1, null)
        // 앞에 남은 사람 수
        val numberInFrontOfMe = waitingService.getNumberInFrontOfMe(waiting.myWaitingNumber, waiting.restaurant.id!!)
        val myWaitingDto = MyWaitingInfoDto(waiting, numberInFrontOfMe)
        return ResultDto(1, myWaitingDto)
    }

    /**
     * 장바구니 리스트 조회
     */
    @PostMapping("/api/customer/{customerId}/baskets")
    fun getBasketList(@PathVariable customerId: Long): ResultDto<BasketListResultDto> {
        val customer = customerService.findCustomerWithBasketWithFood(customerId)
        var restaurantName: String? = null

        customer.basketKey?.let {
            restaurantName = restaurantService.findRestaurant(it).restaurantName
        }

        val basketResponseDtos = customer.baskets.map { BasketFood(it) }
        val basketListResultDto = BasketListResultDto(restaurantName, basketResponseDtos)
        return ResultDto(1, basketListResultDto)
    }

    /**
     * 내 쿠폰 리스트 조회
     */
    @PostMapping("/api/customer/{customerId}/my_coupons")
    fun getMyCouponList(@PathVariable customerId: Long): ResultDto<List<CouponDto>> {
        val myCoupons = customerService.findMyCoupon(customerId)
        val couponDtos = myCoupons.map { CouponDto(it) }
        return ResultDto(couponDtos.size, couponDtos)
    }

    /**
     * 매장 북마크(찜)
     */
    @PostMapping("/api/customer/{customerId}/bookmark")
    fun restaurantBookmark(@PathVariable customerId: Long,
                           @RequestParam(name = "restaurant_id") restaurantId: Long): ResultDto<Long> {
        // 중복 체크
        val bExist: Boolean = customerService.isExistedBookmark(customerId, restaurantId)
        // 이미 북마크 되어 있음
        return if (bExist) {
            ResultDto(1, null)
        } else {
            val bookmarkId = customerService.addBookmark(customerId, restaurantId)
            ResultDto(1, bookmarkId)
        }
    }

    /**
     * 매장 북마크(찜) 해제
     */
    @DeleteMapping("/api/customer/bookmark/{bookmarkId}")
    fun deleteBookmark(@PathVariable bookmarkId: Long): ResultDto<Boolean> {
        customerService.removeBookmark(bookmarkId)
        return ResultDto(1, true)
    }

    /**
     * 내 북마크 불러오기
     */
    @GetMapping("/api/customer/{customerId}/bookmarks")
    fun getBookmarkList(@PathVariable customerId: Long): ResultDto<List<BookmarkPreviewDto>> {
        val bookmarks = customerService.findAllBookmarkWithRestWithRepresentative(customerId)
        val previewDtos = bookmarks.map { BookmarkPreviewDto(it) }
        return ResultDto(previewDtos.size, previewDtos)
    }

    /**
     * 내 주문 내역(진행 중) 리스트 불러오기
     */
    @GetMapping("/api/customer/{customerId}/orders/ongoing")
    fun getAllMyOngoingOrders(@PathVariable customerId: Long): ResultDto<List<OrderPreviewWithRestSimpleDto>> {
        val ongoingOrders = customerService.findAllMyOngoingOrders(customerId)
        val historyDtos = ongoingOrders.map { OrderPreviewWithRestSimpleDto(it) }
        return ResultDto(historyDtos.size, historyDtos)
    }

    /**
     * 내 주문 내역(완료) 리스트 조회
     */
    @GetMapping("/api/customer/{customerId}/orders/finished")
    fun getAllMyFinishedOrders(@PathVariable customerId: Long): ResultDto<List<OrderPreviewWithRestSimpleDto>> {
        val finishedOrders = customerService.findAllMyFinishedOrders(customerId)
        val historyDtos = finishedOrders.map { OrderPreviewWithRestSimpleDto(it) }
        return ResultDto(historyDtos.size, historyDtos)
    }

    /**
     * 내 주문 상세정보
     */
    @GetMapping("/api/customer/order/{orderId}/detail")
    fun getOrderDetail(@PathVariable orderId: Long): ResultDto<OrderDetailDto> {
        val order = orderService.findOrderWithRestWithOrderFoodsWithFood(orderId)
        val orderDetailDto = OrderDetailDto(order)
        return ResultDto(1, orderDetailDto)
    }

    /**
     * 최근 주문 매장 리스트 조회
     */
    @GetMapping("/api/customer/{customerId}/orders/recent")
    fun getRecentOrdersRestaurant(@PathVariable customerId: Long): ResultDto<List<RecentOrderRestaurantDto>> {
        val restaurantDtos = customerService.findRecentOrdersWithRestaurant(customerId, 10)
        return ResultDto(restaurantDtos.size, restaurantDtos)
    }
}